### DISCUSSION TITLE
Estimation bias while working with navego_example_real_ekinox

### DISCUSSION THREAD

**Discussion by Boooooots** [rodralez/NaveGo#74] [Q&A] [answered]
Dear team of NaveGo,
    Thanks to excellent code like naveGo, I am learning it and using it in my project.  I encountered problems with estimation bias while working with navego_example_real_ekinox, which I didn't understand.  Why gyro bias in the state quantity is a variable waveform that fluctuates up to 0.1rad/s.  As I understand it, gyro bias should converge to a constant at the beginning of operation and fluctuate slightly around that constant.  
In addition, I tried to modify the imu.gb_std, and the simulated KF BIAS GYRO X/Y/Z ESTIMATION remains unchanged, which makes me very confused.  

![gyro bias](https://user-images.githubusercontent.com/51846089/142985645-542b9ce0-5be8-41f2-b107-5f720dddeee7.PNG)

In addition, can the all state quantities be set to 0 (including gyro bias and acceleration bias) before the input of kalman filter, and then set to 0 during state correction.
Gyro bias and acceleration bias are accumulated separately.  Like this:  
Kf. Xp (1:15) = 0.0;  % states 1 to 15 are forced to be zero  
.  
.  
% Biases estimation  
gb_dyn += kf.xp(10:12);   % accumulation  
ab_dyn += kf.xp(13:15);  

Thank you and look forward to your reply.

**Comment by rodralez** [comment, marked as answer]
Hi, 

The bias is modeled as having a constant part (static, xb_sta) and a changing part (dynamic, xb_dyn), which is quite convenient when dealing with noisy IMUs.

The imu.gb_std and imu.ab_std, standard deviations, are used for simulating the inertial sensors. You should modify imu.arw and imu.vrw, respectively, to set the level of noise of your inertial sensors to adjust the performance of your navigation system.

The kf.xp(1:9) states are the only error states (Δx) and are set to zero before each Kalman filter iteration. The rest of the states, kf.xp(10:15), are not error states since they are used for both biases estimation and correction. 

Kind regards.

**Comment by Boooooots** [comment]
Thanks a lot,
Sorry, I would like to express the gb_sta (gyro static bais), but I wrote gb_std by mistake.  
I read the Figure 9 that the Gyro Bais reached 0.1rad/s. The reason for my confusion is that I think the Bais is too big.  
When I look at the source code for plot relevant parts, you should have meant deg/s, but log is marked rad/s. (maybe this is a minor bug)  
![1](https://user-images.githubusercontent.com/51846089/145362253-51b58a3e-90cb-4bac-97d0-ef8108825267.PNG)

 
The original gb_sta was [-0.2451 0.0123-0.0222]*E-3. I tried to set gb_sta to [0 0 0] and compared gb_dyn under the two conditions.  
It is found that gb_dyn obtained when gb_sta is set to [0 0 0] is indeed smaller than when gb_sta is [-0.2451 0.0123-0.0222]*E-3.  But the minus is not equal to [-0.2451 0.0123-0.0222]*e-3. [such as: 1.8025 e-4 + (0.2451 e-3) is not equal to (3.4612 e-5) + 0] . I think gb_dyn add gb_sta should remain unchanged.  I wonder if this understanding is correct?  
 
![2](https://user-images.githubusercontent.com/51846089/145362394-92a7a514-dac7-4e23-9afd-5ccd263f622d.PNG)

 
In the case of large bias fluctuation, if entering the tunnel at this time, pure INS update will bring large drift in the case of gyro bias inaccurate.  
 
 
I have checked some papers about ESKF and found that bias is indeed used as error, after iteration ，the bias is reset to zero, and then bias is accumulated.  (paper : Quaternion kinematics for the error-state Kalman filter.pdf)
I don't understand the difference and connection between the two approaches.  Does it say that there is less bias fluctuation?  
 
![4](https://user-images.githubusercontent.com/51846089/145362623-3e3902b1-db10-47bd-a9a1-85fac34739e9.PNG)

 
One problem with the specific force equation, I read some reference books that the Coriolis acceleration is the cross product of the transfer angular rate add the earth's rotation and velocity.  
This is the opposite of your code.  You are the cross product of the velocity and the transfer Angle add the earth's rotation.  
The opposite cross product leads to different velocity predictions.
![3](https://user-images.githubusercontent.com/51846089/145362863-437118c1-9b81-4e04-88b4-08611f741510.PNG)
![5](https://user-images.githubusercontent.com/51846089/145362893-b776e011-4424-4ebc-b9de-4e66b2908641.PNG)

Thank you and look forward to your reply.
Kind regards.

**Reply by rodralez** [reply]
> Thanks a lot, Sorry, I would like to express the gb_sta (gyro static bais), but I wrote gb_std by mistake. I read the Figure 9 that the Gyro Bais reached 0.1rad/s. The reason for my confusion is that I think the Bais is too big. When I look at the source code for plot relevant parts, you should have meant deg/s, but log is marked rad/s. (maybe this is a minor bug) ![1](https://user-images.githubusercontent.com/51846089/145362253-51b58a3e-90cb-4bac-97d0-ef8108825267.PNG)

Hi, yes this is a bug. I just corrected. Thank you very much for pointing it out. 

> The original gb_sta was [-0.2451 0.0123-0.0222]*E-3. I tried to set gb_sta to [0 0 0] and compared gb_dyn under the two conditions. It is found that gb_dyn obtained when gb_sta is set to [0 0 0] is indeed smaller than when gb_sta is [-0.2451 0.0123-0.0222]*E-3. But the minus is not equal to [-0.2451 0.0123-0.0222]*e-3. [such as: 1.8025 e-4 + (0.2451 e-3) is not equal to (3.4612 e-5) + 0] . I think gb_dyn add gb_sta should remain unchanged. I wonder if this understanding is correct?
> 
> ![2](https://user-images.githubusercontent.com/51846089/145362394-92a7a514-dac7-4e23-9afd-5ccd263f622d.PNG)
> 
> In the case of large bias fluctuation, if entering the tunnel at this time, pure INS update will bring large drift in the case of gyro bias inaccurate.

If you set to zero the static bias, for your particular sensors and configuration, it seems that the Kalman filter cannot estimate correctly the static bias. You may add extra states to estimate the static biases. But a better idea is to provide NaveGo an estimation of static biases. 

> I have checked some papers about ESKF and found that bias is indeed used as error, after iteration ，the bias is reset to zero, and then bias is accumulated. (paper : Quaternion kinematics for the error-state Kalman filter.pdf) I don't understand the difference and connection between the two approaches. Does it say that there is less bias fluctuation?
> 
> ![4](https://user-images.githubusercontent.com/51846089/145362623-3e3902b1-db10-47bd-a9a1-85fac34739e9.PNG)

There are several models for INS/GNSS systems. NaveGo is mostly based on Groves's book, 2nd edition. We are always open to implementing better models but so far this is the best model we have tried. 

> One problem with the specific force equation, I read some reference books that the Coriolis acceleration is the cross product of the transfer angular rate add the earth's rotation and velocity. This is the opposite of your code. You are the cross product of the velocity and the transfer Angle add the earth's rotation. The opposite cross product leads to different velocity predictions. ![3](https://user-images.githubusercontent.com/51846089/145362863-437118c1-9b81-4e04-88b4-08611f741510.PNG) ![5](https://user-images.githubusercontent.com/51846089/145362893-b776e011-4424-4ebc-b9de-4e66b2908641.PNG)

Thank you very much for pointing out this bug. I took this formula from Titterton's book, but I misunderstanded the mathematical symbol 'x' which is a vector cross product, not a multiplication by a skew matrix.  Nevertheless, the Coriolis effect only has an influence when using high-end IMUs.

Kind regards.

**Reply by Boooooots** [reply]
Thanks a lot, 
I will continue to learn and use Navego, and convert MATLAB to C language to implement it on STM32F767 platform.
I will Pay a close attention to this Navego project.

Kind regards.

**Reply by na851998** [reply]
In Grove's matlab code, I noticed that he also implemented loosely-coupled integration like mentioned above:
Kf. Xp (1:15) = 0.0; % states 1 to 15 are forced to be zero
.
.
% Biases estimation
gb_dyn += kf.xp(10:12); % accumulation
ab_dyn += kf.xp(13:15);

So maybe the way you estimate bias:
gb_dyn = -kf.xp(10:12); 
ab_dyn = -kf.xp(13:15);
Were these equations from other books or papers?

Thank you and look forward to your reply.
Kind regards.

**Reply by rodralez** [reply]
> In Grove's matlab code, I noticed that he also implemented loosely-coupled integration like mentioned above: 

> Kf. Xp (1:15) = 0.0; % states 1 to 15 are forced to be zero . . 
> % Biases estimation 
> gb_dyn += kf.xp(10:12); % accumulation 
> ab_dyn += kf.xp(13:15);

Biases estimation states are not error states in NaveGo. They estimate the total value of each inertial sensor bias. So, they cannot be initialized as zero. 

Please, can you point out the MATLAB file from Grove's exercises where you took the code above from?

> So maybe the way you estimate bias: 
> gb_dyn = -kf.xp(10:12); 
> ab_dyn = -kf.xp(13:15); 

> Were these equations from other books or papers?

I took them from Groves, so maybe I have to double check if they are right.

**Reply by na851998** [reply]
In Groves code, static bias and dynamic bias are considered as one bias which needs to be estimated (initilized as 0). This bias is updated in **LC_KF_Epoch.m** everytime GNSS measurements are available. In **LC_KF_Epoch.m**, all states are reset to 0 in Step 3, get updated in Step 9 and finally, IMU biases are accumulated at the last line

Kind regards.

**Reply by rodralez** [reply]
Hi, I checked that inertial biases states are error states in Kalman filter. They are the errors between both truth and noisy inertial measurements, as shown in Eq. 4.18, Groves' book.

So, you were right. Thank you very much for pointing out this bug. 

This comment is related to issue #129.

**Reply by na851998** [reply]
Dear mr.rodralez,
I wondered if inertial biases states are error states, shouldn't it be added to gb_dyn or ab_dyn to correct the estimate

Kind regards.
